src/components/ModelingMachineProvider.tsx

```diff
@@ -215,154 +222,175 @@ export const ModelingMachineProvider = ({
               }
-            } else if (event.data.type === 'timeoutEnd') {
+            } else if (_event.data.type === 'timeoutEnd') {
               const copy = { ...segmentHoverMap }
-              delete copy[event.data.pathToNodeString]
+              delete copy[_event.data.pathToNodeString]
               return copy
             }
             return {}
           },
         }),
         'Set Segment Overlays': assign({
-          segmentOverlays: ({ segmentOverlays }, { data }) => {
-            if (data.type === 'set-many') return data.overlays
-            if (data.type === 'set-one')
+          segmentOverlays: ({ context: { segmentOverlays }, event }) => {
+            const _event = event as Extract<
+              ModelingMachineEvent,
+              { type: 'Set Segment Overlays' }
+            >
+            if (_event.data.type === 'set-many') return _event.data.overlays
+            if (_event.data.type === 'set-one')
               return {
                 ...segmentOverlays,
-                [data.pathToNodeString]: data.seg,
+                [_event.data.pathToNodeString]: _event.data.seg,
               }
-            if (data.type === 'delete-one') {
+            if (_event.data.type === 'delete-one') {
               const copy = { ...segmentOverlays }
-              delete copy[data.pathToNodeString]
+              delete copy[_event.data.pathToNodeString]
               return copy
             }
             // data.type === 'clear'
             return {}
           },
         }),
-        'Set sketchDetails': assign(({ sketchDetails }, event) =>
+        'Set sketchDetails': assign(({ context: { sketchDetails }, event }) =>
           sketchDetails
             ? {
                 sketchDetails: {
                   ...sketchDetails,
-                  sketchPathToNode: event.data,
+                  sketchPathToNode: (
+                    event as Extract<
+                      ModelingMachineEvent,
+                      { type: 'Delete segment' }
+                    >
+                  ).data,
                 },
               }
             : {}
         ),
-        'Set selection': assign(({ selectionRanges, sketchDetails }, event) => {
-          const setSelections = event.data as SetSelections // this was needed for ts after adding 'Set selection' action to on done modal events
-          const dispatchSelection = (selection?: EditorSelection) => {
-            if (!selection) return // TODO less of hack for the below please
-            if (!editorManager.editorView) return
-            setTimeout(() => {
+        'Set selection': assign(
+          ({ context: { selectionRanges, sketchDetails }, event }) => {
+            const _event = event as Extract<
+              ModelingMachineEvent,
+              { type: 'Set selection' }
+            >
+            const setSelections = _event.data // this was needed for ts after adding 'Set selection' action to on done modal events
```

Comment:
@farskid one problem I'm running into is that our constraint actors (such as `Get horizontal info`) call this action `onDone`, and when they do their payload is in `_event.output`, not `_event.data`. Do we need TS types to spell out that there is an `output` property of the same type as `data` would be?

Reply:
I believe I've fixed this with my latest commits, maybe let me know if how I solved it [here](https://github.com/KittyCAD/modeling-app/pull/3735/commits/66c6e0635204d453cdf7bbb8a98f0132defd43ef) is an okay practice?

Reply:
https://github.com/KittyCAD/modeling-app/pull/3735#discussion_r1744045159